storage/remote/store.go

```diff
@@ -0,0 +1,144 @@
+	"github.com/onflow/flow-emulator/storage/sqlite"
+	exeState "github.com/onflow/flow-go/engine/execution/state"
+	"github.com/onflow/flow-go/fvm/state"
+	"github.com/onflow/flow-go/fvm/storage/snapshot"
+	"github.com/onflow/flow-go/ledger/common/pathfinder"
+	"github.com/onflow/flow-go/ledger/complete"
+	flowgo "github.com/onflow/flow-go/model/flow"
+	"github.com/pkg/errors"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/credentials/insecure"
+)
+
+type Store struct {
+	*sqlite.Store
+	client archive.APIClient
+}
+
+func New(chainID flowgo.ChainID) (*Store, error) {
+	archiveHosts := map[flowgo.ChainID]string{
+		flowgo.Mainnet: "archive.mainnet.nodes.onflow.org:9000",
+		flowgo.Testnet: "archive.testnet.nodes.onflow.org:9000",
+	}
+
+	host, ok := archiveHosts[chainID]
+	if !ok {
+		return nil, fmt.Errorf("chain %s not supported with remote store", chainID.String())
+	}
+
+	conn, err := grpc.Dial(
+		host,
+		grpc.WithTransportCredentials(insecure.NewCredentials()),
+	)
+	if err != nil {
+		return nil, fmt.Errorf("could not connect to archive node: %w", err)
+	}
+
+	memorySql, err := sqlite.New(sqlite.InMemory)
+	if err != nil {
+		return nil, err
+	}
+
+	store := &Store{
+		client: archive.NewAPIClient(conn),
+		Store:  memorySql,
+	}
+
+	store.DataGetter = store
+	store.DataSetter = store
+	store.KeyGenerator = &storage.DefaultKeyGenerator{}
+
+	return store, nil
+}
+
+func (s *Store) BlockByID(ctx context.Context, blockID flowgo.Identifier) (*flowgo.Block, error) {
+	// todo use local storage first as a cache
+
+	heightRes, err := s.client.GetHeightForBlock(ctx, &archive.GetHeightForBlockRequest{BlockID: blockID[:]})
+	if err != nil {
+		return nil, err
+	}
+
+	return s.BlockByHeight(ctx, heightRes.Height)
+}
+
+func (s *Store) LatestBlock(ctx context.Context) (flowgo.Block, error) {
+	heightRes, err := s.client.GetLast(ctx, &archive.GetLastRequest{})
+	if err != nil {
+		return flowgo.Block{}, err
+	}
+
+	block, err := s.BlockByHeight(ctx, heightRes.Height)
+	if err != nil {
+		return flowgo.Block{}, err
+	}
+
+	return *block, nil
+}
+
+func (s *Store) BlockByHeight(ctx context.Context, height uint64) (*flowgo.Block, error) {
+	// todo use local storage first as a cache
+
+	blockRes, err := s.client.GetHeader(ctx, &archive.GetHeaderRequest{Height: height})
+	if err != nil {
+		return nil, err
+	}
+
+	var header flowgo.Header
+	err = zbor.NewCodec().Unmarshal(blockRes.Data, &header)
+	if err != nil {
+		return nil, err
+	}
+
+	payload := flowgo.EmptyPayload()
+	return &flowgo.Block{
+		Payload: &payload,
+		Header:  &header,
+	}, nil
+}
+
+func (s *Store) LedgerByHeight(
+	ctx context.Context,
+	blockHeight uint64,
+) state.StorageSnapshot {
+	_ = s.SetBlockHeight(blockHeight)
+
+	return snapshot.NewReadFuncStorageSnapshot(func(id flowgo.RegisterID) (flowgo.RegisterValue, error) {
+		// first try to see if we have local stored ledger
+		value, err := s.DefaultStore.GetBytesAtVersion(ctx, "ledger", []byte(id.String()), blockHeight)
```

Comment:
Maybe refactor the magic `"ledger"` value into a constant.

Where is it used for writing?

Reply:
The writing part is in another package and I don't like exposing it, but yeah I can make it a constant.